src/org/intellij/erlang/completion/ErlangAppCompletionContributor.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2012-2015 Sergey Ignatov
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.intellij.erlang.completion;
+
+import com.intellij.codeInsight.completion.*;
+import com.intellij.codeInsight.lookup.LookupElementBuilder;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.intellij.util.ProcessingContext;
+import com.intellij.util.containers.ContainerUtil;
+import org.intellij.erlang.ErlangFileType;
+import org.intellij.erlang.psi.ErlangFile;
+import org.intellij.erlang.psi.ErlangListExpression;
+import org.intellij.erlang.psi.ErlangTupleExpression;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Set;
+
+import static com.intellij.patterns.PlatformPatterns.psiElement;
+import static com.intellij.patterns.StandardPatterns.instanceOf;
+
+public class ErlangAppCompletionContributor extends CompletionContributor {
+  // source: "FILE SYNTAX" section of http://www.erlang.org/doc/man/app.html
+  public static final Set<String> KEYWORDS = ContainerUtil.newHashSet("description", "id", "vsn", "modules", "maxP",
+                                                                      "maxT", "registered", "included_applications",
+                                                                      "applications", "env", "mod", "start_phases",
+                                                                      "runtime_dependencies");
+
+  @Override
+  public void beforeCompletion(@NotNull CompletionInitializationContext context) {
+    PsiFile file = context.getFile();
+    if (file.getFileType() != ErlangFileType.APP) return;
+    int startOffset = context.getStartOffset();
+    PsiElement elementAt = file.findElementAt(startOffset);
+    PsiElement parent = elementAt != null ? elementAt.getParent() : null;
+    if (parent instanceof ErlangTupleExpression) {
+      context.setDummyIdentifier("a");
+    }
```

Comment:
Now - yes. Because without dummy identifier first completion ("application") does not work at all (because PSI structure is broken and there is no tuple), and second completion (all other keywords) has different levels of parents with or without text.

With dummy identifier all is ok.